### ISSUE TITLE
Return values

### ISSUE THREAD

**Issue by rwaldron** [awtk/Array.prototype.sample#7]
This won't get far with the current return values. I read the Ruby spec, and it's indefensible. Here's my recommendation: 

- Always return an array, in every case
- If source array is empty, return a new empty array
- If called with no arguments, or an argument that equals 1, return a new array containing one randomly chosen entry from the source array
- If called with an argument `n`, return a new array containing `n` randomly chosen entries from the source array.

(This also guarantees that all array methods can be chained)

**Comment by jmeas** [comment]
These suggestions make a lot of sense to me @rwaldron . I was not a fan of `([]).sample() === null` : P

**Comment by rwaldron** [comment]
cc @brianloveswords

**Comment by brianloveswords** [comment]
I agree with this assessment and these suggestions.

**Comment by tkellen** [comment]
@rwaldron 

```
irb(main):001:0> [].sample
=> nil
```

?

**Comment by tkellen** [comment]
```
irb(main):005:0> [1,2,3].sample
=> 3
```

**Comment by tkellen** [comment]
@rwaldron I disagree with point one. I also think your reading of the ruby docs was incorrect (see my previous comment). Also, lodash has the same same behavior as Ruby (returns undefined on an empty array and a single element when requesting a single element).

That said, if there is consensus that the return value should always be an array, I don't feel strongly enough to object. Mostly I'm excited about the possibility of this landing at all!

**Comment by tkellen** [comment]
/me wakes up and realizes the definition of indefensible.

Disregard my comment about the incorrect reading of the spec!

**Comment by rwaldron** [comment]
@tkellen I posted this recommendation after carefully reviewing the docs and even trying the same thing in irb. The Ruby behavior that I described as "indefensible" is the behavior you've just illustrated. The Ruby semantics are messy and JS shouldn't copy the broken parts of the feature. Programs should be able to call `a.sample()` and safely call any other array method from the return value of that call, without requiring some null check first. A concrete case: If I used `a.sample()` with some running list of values that were derived from async hardware reads, I absolutely would not want a call that occurs before any reads have been delivered (or after clearing the received values) to result in a non-array value. 

Example: 

```js
class Sensor extends GPIO {
  constructor() {
    super();
    this.received = [];
    // GPIO base class emits "data" events asynchronously, so the first
    // won't occur until at least the next execution turn.
    this.on("data", value => this.received.push(value));
  }
  sampledAvg(size) {
    size = size || 3;
    return this.received.sample(size).reduce((a, b) => a + b, 0) / size;
  }
}

let sensor = new Sensor();

// With my recommendation
sensor.sampledAvg(3); // 0 (because the array is empty, and 0 is the initial value provided to reduce

// With a null return
sensor.sampledAvg(3); // throws because I'm trying to call reduce on null :(
```
